javaslang/src/main/java/javaslang/collection/Collections.java

```diff
@@ -110,25 +110,18 @@ static int hash(Iterable<?> iterable) {
         return hashCode;
     }
 
-    static <T, U, C extends Iterable<U>, R extends Traversable<U>> R scanLeft(Iterable<? extends T> elements,
-                                                                              U zero, BiFunction<? super U, ? super T, ? extends U> operation,
-                                                                              C cumulativeResult, BiFunction<C, U, C> combiner,
-                                                                              Function<C, R> finisher) {
-        U acc = zero;
-        cumulativeResult = combiner.apply(cumulativeResult, acc);
-        for (T a : elements) {
-            acc = operation.apply(acc, a);
-            cumulativeResult = combiner.apply(cumulativeResult, acc);
-        }
-        return finisher.apply(cumulativeResult);
+    static <T, U, R extends Traversable<U>> R scanLeft(Traversable<? extends T> traversable,
+                                                       U zero, BiFunction<? super U, ? super T, ? extends U> operation, Function<Iterator<U>, R> finisher) {
+
+        final Iterator<U> iterator = traversable.iterator().scanLeft(zero, operation);
+        return finisher.apply(iterator);
     }
 
-    static <T, U, C extends Iterable<U>, R extends Traversable<U>> R scanRight(Iterable<? extends T> elements,
-                                                                               U zero, BiFunction<? super T, ? super U, ? extends U> operation,
-                                                                               C cumulativeResult, BiFunction<C, U, C> combiner,
-                                                                               Function<C, R> finisher) {
-        final Iterator<? extends T> reversedElements = reverseIterator(elements);
-        return scanLeft(reversedElements, zero, (u, t) -> operation.apply(t, u), cumulativeResult, combiner, finisher);
+    static <T, U, R extends Traversable<U>> R scanRight(Traversable<? extends T> traversable,
+                                                        U zero, BiFunction<? super T, ? super U, ? extends U> operation, Function<Iterator<U>, R> finisher) {
+
+        final Iterator<? extends T> reversedElements = reverseIterator(traversable.iterator());
```

Comment:
Why do you need the `.iterator()` exactly, `reverseIterator(traversable)` should work also?